lib/compat/wordpress-6.8/site-editor.php

```diff
@@ -0,0 +1,124 @@
+		return add_query_arg( array( 'p' => '/pattern' ), remove_query_arg( 'postType' ) );
+	}
+
+	if ( isset( $_REQUEST['postType'] ) && 'wp_block' === $_REQUEST['postType'] && ! empty( $_REQUEST['postId'] ) ) {
+		return add_query_arg( array( 'p' => '/wp_block/' . $_REQUEST['postId'] ), remove_query_arg( array( 'postType', 'postId' ) ) );
+	}
+
+	if ( isset( $_REQUEST['postType'] ) && 'wp_template_part' === $_REQUEST['postType'] && ( empty( $_REQUEST['canvas'] ) || empty( $_REQUEST['postId'] ) ) ) {
+		return add_query_arg( array( 'p' => '/pattern' ) );
+	}
+
+	if ( isset( $_REQUEST['postType'] ) && 'wp_template_part' === $_REQUEST['postType'] && ! empty( $_REQUEST['postId'] ) ) {
+		return add_query_arg( array( 'p' => '/wp_template_part/' . $_REQUEST['postId'] ), remove_query_arg( array( 'postType', 'postId' ) ) );
+	}
+
+	// The following redirects are for backward compatibility with the old site editor URLs.
+	if ( isset( $_REQUEST['path'] ) && '/wp_template_part/all' === $_REQUEST['path'] ) {
+		return add_query_arg(
+			array(
+				'p'        => '/pattern',
+				'postType' => 'wp_template_part',
+			),
+			remove_query_arg( 'path' )
+		);
+	}
+
+	if ( isset( $_REQUEST['path'] ) && '/page' === $_REQUEST['path'] ) {
+		return add_query_arg( array( 'p' => '/page' ), remove_query_arg( 'path' ) );
+	}
+
+	if ( isset( $_REQUEST['path'] ) && '/wp_template' === $_REQUEST['path'] ) {
+		return add_query_arg( array( 'p' => '/template' ), remove_query_arg( 'path' ) );
+	}
+
+	if ( isset( $_REQUEST['path'] ) && '/patterns' === $_REQUEST['path'] ) {
+		return add_query_arg( array( 'p' => '/pattern' ), remove_query_arg( 'path' ) );
+	}
+
+	if ( isset( $_REQUEST['path'] ) && '/navigation' === $_REQUEST['path'] ) {
+		return add_query_arg( array( 'p' => '/navigation' ), remove_query_arg( 'path' ) );
+	}
+
+	return add_query_arg( array( 'p' => '/' ) );
+}
+
+function gutenberg_redirect_site_editor_deprecated_urls() {
+	$redirection = gutenberg_get_site_editor_redirection();
+	if ( false !== $redirection ) {
+		wp_redirect( $redirection, 301 );
+		exit;
+	}
+}
+add_action( 'admin_init', 'gutenberg_redirect_site_editor_deprecated_urls' );
+
+/**
+ * Filter the `wp_die_handler` to allow access to the Site Editor's new pages page
+ * for Classic themes.
+ *
+ * site-editor.php's access is forbidden for hybrid/classic themes and only allowed with some very special query args (some very special pages like template parts...).
+ * The only way to disable this protection since we're changing the urls in Gutenberg is to override the wp_die_handler.
+ *
+ * @param callable $default_handler The default handler.
+ * @return callable The default handler or a custom handler.
+ */
+function gutenberg_styles_wp_die_handler( $default_handler ) {
+	if ( ! wp_is_block_theme() && str_contains( $_SERVER['REQUEST_URI'], 'site-editor.php' ) && isset( $_GET['p'] ) ) {
```

Comment:
Why do we need `&& isset( $_GET['p'] )`? This means it's not possible to access the site editor root, which I need for #66851